src/lava/magma/core/learning/learning_rule.py

```diff
@@ -0,0 +1,567 @@
+    ) -> None:
+        # dict of string learning rules
+        str_learning_rules = {
+            str_symbols.DW: dw,
+            str_symbols.DD: dd,
+            str_symbols.DT: dt,
+        }
+
+        # dict of string learning rules that were provided only
+        active_str_learning_rules = {
+            key: str_learning_rule
+            for key, str_learning_rule in str_learning_rules.items()
+            if str_learning_rule is not None
+        }
+
+        # validate that at least one learning rule was provided
+        self._validate_at_least_one_learning_rule(active_str_learning_rules)
+        # validate that same k was used throughout all learning rules in case
+        # a uk dependency is used
+        self._decimate_exponent = self._validate_uk(active_str_learning_rules)
+
+        # extract and validate x trace impulses and taus
+        self._x1_impulse = self._validate_impulse(x1_impulse)
+        self._x2_impulse = self._validate_impulse(x2_impulse)
+        self._x1_tau = self._validate_tau(x1_tau)
+        self._x2_tau = self._validate_tau(x2_tau)
+
+        # extract and validate y1 trace impulse and tau
+        self._y1_impulse = self._validate_impulse(y1_impulse)
+        self._y1_tau = self._validate_tau(y1_tau)
+        self._y2_impulse = self._validate_impulse(y2_impulse)
+        self._y2_tau = self._validate_tau(y2_tau)
+        self._y3_impulse = self._validate_impulse(y3_impulse)
+        self._y3_tau = self._validate_tau(y3_tau)
+
+        # extract and validate t_epoch (epoch length)
+        self._t_epoch = self._validate_t_epoch(t_epoch)
+
+        # set seed or generate by time
+        self._rng_seed = rng_seed if rng_seed is not None else int(time.time())
+
+        # generate ProductSeries for all learning rules that were provided in
+        # string format
+        self._active_product_series = {
+            key: self._generate_product_series_from_string(
+                key, str_learning_rule
+            )
+            for key, str_learning_rule in active_str_learning_rules.items()
+        }
+
+        # set attribute for each of the active ProductSeries
+        self._dw = self._dd = self._dt = None
+        for key, product_series in self._active_product_series.items():
+            setattr(self, f"_{key}", product_series)
+
+        # active_traces: set of traces used by all learning rules that
+        # were specified.
+        # active_traces_per_dependency: dict with dependencies as keys and
+        # the set of all traces appearing with them in the specified
+        # learning rules as values
+        (
+            self._active_traces,
+            self._active_traces_per_dependency,
+        ) = self._get_active_traces_from_active_product_series()
+
+    @property
+    def x1_impulse(self) -> float:
+        """Get the impulse value for x1 trace.
+
+        Returns
+        ----------
+        x1_impulse : float
+            Impulse value for x1 trace.
+        """
```

Comment:
I am questioning whether the docstrings for all of these decorators are really valuable to users. They give little more information than the code itself and hinder developers in quickly browsing the code. I would be in favor of removing them all and making the code more expressive instead.
Probably not something we should be doing now. Just a general thing to keep in mind.

Reply:
agreed, we will try to refactor this at a later timepoint.